### DISCUSSION TITLE
The running mechanism of the program after configuring s3 in storagedriver

### DISCUSSION THREAD

**Discussion by balabalabutton** [distribution/distribution#4497] [Q&A]
In the configuration docs, there is a sentence as follows:  「Note When S3 storage is used, registry redirects to a pre-signed URL for the requested object. By not having to serve the stored images directly, this allows running registry on hosts with lightweight resource requirements.」

storageDriver.redirect is also started by default，

So, my understanding is that after distribution stores the image into s3 object storage, if cri such as docker/containerd requests distribution to obtain the image, distribution will parse the address of the image layers data in s3, and then notify cri of the address, and cri will download it from s3 itself , rather than the distribution itself downloading the layers and then transmitting them to the cri client.

But when I asked chatGPT or other AI at the beginning, the answer was that distribution pulls the data and then transmits it to the cri client.

So, I'm confused......

So, the logic of distribution should be to tell cri the download address of the layers to be pulled in s3, right?

**Comment by milosgajdos** [comment]
**Please do not rely on everything ChatGPT tells you! The source code is freely available, read it.**

The actual behaviour depends on the configuration; but, yes you can configure distribution so it responds with a presigned S3 URL and then the client can just follow it and download the blobs:

https://github.com/distribution/distribution/blob/c427f845035d28faab65a9218407293bd917e7d5/registry/storage/driver/s3-aws/s3.go#L1105
